src/psyclone/dynamo0p3.py

```diff
@@ -2904,40 +2913,34 @@ class DynScalarArgs(DynCollection):
     :type node: :py:class:`psyclone.dynamo0p3.DynKern` or \
                 :py:class:`psyclone.dynamo0p3.DynInvoke`
 
-    :raises InternalError: if an unrecognised intrinsic type of the scalar \
-                           argument is encountered.
-
     '''
     def __init__(self, node):
         super(DynScalarArgs, self).__init__(node)
 
         # Create lists of real and integer scalar arguments
         self._real_scalar_names = {}
         self._int_scalar_names = {}
+        scalar_args = {}
+        # Filter scalar arguments by intent
         for intent in FORTRAN_INTENT_NAMES:
             self._real_scalar_names[intent] = []
             self._int_scalar_names[intent] = []
+            scalar_args[intent] = []
         if self._invoke:
-            for scalar_type in LFRicArgDescriptor.VALID_SCALAR_NAMES:
-                scalar_dict = self._invoke.unique_declns_by_intent(scalar_type)
-                for intent in FORTRAN_INTENT_NAMES:
-                    for arg in scalar_dict[intent]:
-                        declname = arg.declaration_name
-                        if arg.intrinsic_type == "real":
-                            self._real_scalar_names[intent].append(declname)
-                        if arg.intrinsic_type == "integer":
-                            self._int_scalar_names[intent].append(declname)
+            scalar_args = self._invoke.unique_declns_by_intent(
+                LFRicArgDescriptor.VALID_SCALAR_NAMES)
         else:
             for arg in self._calls[0].arguments.args:
-                if arg.is_scalar():
-                    if arg.intrinsic_type == "real":
-                        self._real_scalar_names[arg.intent].append(arg.name)
-                    elif arg.intrinsic_type == "integer":
-                        self._int_scalar_names[arg.intent].append(arg.name)
-                    else:
-                        raise InternalError(
-                            "Unsupported scalar intrinsic type '{0}' found "
-                            "in DynScalarArgs.".format(arg.intrinsic_type))
+                if arg.is_scalar:
+                    scalar_args[arg.intent].append(arg)
+        # Separate scalars by their intrinsic types
+        for intent in FORTRAN_INTENT_NAMES:
+            for arg in scalar_args[intent]:
+                declname = arg.declaration_name
+                if arg.intrinsic_type == "real":
+                    self._real_scalar_names[intent].append(declname)
+                if arg.intrinsic_type == "integer":
+                    self._int_scalar_names[intent].append(declname)
```

Comment:
For defensive purposes this should be:

    if real:
        ...
    elif integer:
        ...
    else:
        raise InternalError(...)

(That way, if/when we should extend the list of supported scalar types we'll get a nice error message at this point.)

Reply:
Done. Restored and adapted the `test_dynscalars_err` to `test_dynscalars_stub_err` in `dynamo0p3_stubgen_test.py` for testing the kernel stub. Added `test_dynscalars_call_err` in `dynamo0p3_test.py` to test the kernel call.

Reply:
Please include the valid types in the error message.